File: Android/app/src/main/java/com/spiraclestudios/autoskola/IntersectionCanvas.java
/*
 * Copyright (c) 2015-2016. Spiracle Studios. All Rights Reserved.
 */

package com.spiraclestudios.autoskola;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.AssetManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.util.DisplayMetrics;
import android.widget.ImageView;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import timber.log.Timber;

/**
 * Added by benji on 19/12/2015.
 */
public class IntersectionCanvas extends ImageView {

  private static final String contentPath = "images\\Designer Content\\";

  private Context context;
  private Canvas canvas;
  private Bitmap finalBitmap;
  private Bitmap carImage;
  private List<IntersectionObject> objects = new ArrayList<>();
  private Paint paint;

  public IntersectionCanvas(Context c, AttributeSet attrs) {
    super(c, attrs);
    context = c;
    initCanvas();
    loadAssets();
  }

  // http://developer.android.com/reference/android/view/View.html#onMeasure(int, int)
  /*@Override
  protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        // Try for a width based on our minimum
        int minw = getPaddingLeft() + getPaddingRight() + getSuggestedMinimumWidth();
        int w = resolveSizeAndState(minw, widthMeasureSpec, 1);

        // Whatever the width ends up being, ask for a height that would let the pie
        // get as big as it can
        int minh = MeasureSpec.getSize(w) - (int) mTextWidth + getPaddingBottom() + getPaddingTop();
        int h = resolveSizeAndState(MeasureSpec.getSize(w) - (int) mTextWidth, heightMeasureSpec, 0);

        setMeasuredDimension(w, h);
    }*/

  @Override protected void onSizeChanged(int w, int h, int oldw, int oldh) {
    super.onSizeChanged(w, h, oldw, oldh);
    //initCanvas();
    drawIntersection();
  }

  // TODO: Cache all the images used by this intersection from assets.
  // TODO: Take an param for which intersection json to load the assets from.
  private void loadAssets() {
    BitmapFactory.Options options = new BitmapFactory.Options();
    options.inScaled = false;
    //options.inDensity = DisplayMetrics.DENSITY_HIGH;
    //options.inTargetDensity = res.getDisplayMetrics().densityDpi;

    //carImage = BitmapFactory.decodeResource(getResources(), R.drawable.car, options);

    Bitmap image = null;

    getBitmapFromAsset(context, contentPath + "Roads\\road_1.png");
  }

  @SuppressLint("BinaryOperationInTimber")
  private static Bitmap getBitmapFromAsset(Context context, String filePath) {
    AssetManager assetManager = context.getAssets();

    InputStream inputStream;
    Bitmap bitmap = null;
    try {
      inputStream = assetManager.open(filePath);
      bitmap = BitmapFactory.decodeStream(inputStream);
    } catch (IOException e) {
      Timber.d("Image \"%s\" does not exist.", contentPath + filePath);
    }

    return bitmap;
  }

  private void initCanvas() {
    finalBitmap = Bitmap.createBitmap(960, 540, Bitmap.Config.ARGB_8888);
    canvas = new Canvas(finalBitmap);
    canvas.setDensity(DisplayMetrics.DENSITY_HIGH);
  }

  private void drawObject(Bitmap image, float x, float y, float angle) {
    Matrix matrix = new Matrix();
    float w = image.getWidth();
    float h = image.getHeight();

    matrix.setTranslate(x - w / 2, y - h / 2);
    matrix.postRotate(angle, x, y);
    canvas.drawBitmap(image, matrix, paint);
  }

  private void drawIntersection() {
    Matrix trans = new Matrix();
    float x, y;
    float w = carImage.getWidth();
    float h = carImage.getHeight();
    float canvasW = canvas.getWidth();
    float canvasH = canvas.getHeight();
    paint = new Paint();
    paint.setAntiAlias(true);

    // Paint the background.
    canvas.drawColor(Color.parseColor("#607D8B"));

    // Center of screen, rotated.
    drawObject(carImage, canvasW / 2, canvasH / 2, 45);
    // Bottom right corner.
    drawObject(carImage, canvasW - w / 2, canvasH - h / 2, 0);
    // Bottom right corner, rotated.
    drawObject(carImage, canvasW - h / 2, canvasH - w / 2 - h, 90);

    // Mark the canvas' center point
    x = canvasW / 2;
    y = canvasH / 2;
    float size = 2;
    trans.setTranslate(x, y);
    paint.setColor(Color.GREEN);
    canvas.drawRect(x - size, y - size, x + size, y + size, paint);

    setImageBitmap(finalBitmap);
  }

  public void clearCanvas() {
    canvas.drawColor(Color.parseColor("#607D8B"));
    //finalBitmap.recycle();
    //finalBitmap = null;
    //invalidate();
  }

  // private int rot = 0;
  // private float oldX = 0f;
    /*@Override
    public boolean onTouchEvent(MotionEvent e) {
        if (e.getAction() == MotionEvent.ACTION_DOWN) {
            oldX = e.getX();
            //oldY = e.getY();
        } else if (e.getAction() == MotionEvent.ACTION_MOVE) {
            float delta = (e.getX() - oldX) * 0.7f;
            rot += delta;
            oldX = e.getX();
            //oldY = e.getY();

            drawIntersection();
        }
        return true;
    }*/
}

File: Android/app/src/main/java/com/spiraclestudios/autoskola/IntersectionCanvas.kt
/*
 * Copyright 2015-2016 Spiracle Studios. All Rights Reserved.
 */

package com.spiraclestudios.autoskola

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.util.DisplayMetrics
import android.view.MotionEvent
import android.widget.ImageView

/**
 * Copyright 2015-2016 Spiracle Studios. All Rights Reserved.
 * Original created by benji on 19/12/2015.
 */
class IntersectionCanvas

(c: Context, attrs: AttributeSet) : ImageView(c, attrs) {

    private lateinit var mCanvas: Canvas
    // The image that gets drawn to the screen.
    private lateinit var mFinalBitmap: Bitmap
    private lateinit var mCarImage: Bitmap
    private var mPaint: Paint? = null

    private var rot: Int = 0
    private var oldX: Float = 0f

    // http://developer.android.com/reference/android/view/View.html#onMeasure(int, int)
    /*@Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        // Try for a width based on our minimum
        int minw = getPaddingLeft() + getPaddingRight() + getSuggestedMinimumWidth();
        int w = resolveSizeAndState(minw, widthMeasureSpec, 1);

        // Whatever the width ends up being, ask for a height that would let the pie
        // get as big as it can
        int minh = MeasureSpec.getSize(w) - (int) mTextWidth + getPaddingBottom() + getPaddingTop();
        int h = resolveSizeAndState(MeasureSpec.getSize(w) - (int) mTextWidth, heightMeasureSpec, 0);

        setMeasuredDimension(w, h);
    }*/

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        init()
        drawMeLikeOneOfYourIntersections()
    }

    protected fun init() {
        mFinalBitmap = Bitmap.createBitmap(480, 270, Bitmap.Config.ARGB_8888)
        mCanvas = Canvas(mFinalBitmap)
        mCanvas.density = DisplayMetrics.DENSITY_HIGH

        val options = BitmapFactory.Options()
        options.inScaled = false
        //options.inDensity = DisplayMetrics.DENSITY_HIGH;
        //options.inTargetDensity = res.getDisplayMetrics().densityDpi;

        // TODO: Cache all the images used by this intersection from assets. Move this into importIntersection()
        //InputStream inputStream = assetManager.open(path);
        mCarImage = BitmapFactory.decodeResource(resources, R.drawable.car, options)
    }

    /*protected fun importIntersection() : Intersection
    {
        return
    }*/

    protected fun drawObject(image: Bitmap, x: Float, y: Float, angle: Float) {
        val trans = Matrix()
        val w = image.width.toFloat()
        val h = image.height.toFloat()

        trans.setTranslate(x - w / 2, y - h / 2)
        trans.postRotate(angle, x, y)
        mCanvas.drawBitmap(image, trans, mPaint)
    }

    protected fun drawMeLikeOneOfYourIntersections() {
        val trans = Matrix()
        val x: Float
        val y: Float
        val w = mCarImage.width.toFloat()
        val h = mCarImage.height.toFloat()
        val canvasW = mCanvas.width.toFloat()
        val canvasH = mCanvas.height.toFloat()
        val mPaint = Paint()
        mPaint.isAntiAlias = true

        // Clear screen
        mCanvas.drawColor(Color.parseColor("#e5e5e5"))

        // Center of screen, rotated
        drawObject(mCarImage, canvasW / 2, canvasH / 2, rot.toFloat())
        // Bottom right corner
        drawObject(mCarImage, canvasW - w / 2, canvasH - h / 2, 0f)
        // Bottom right corner, rotated
        mPaint.setColorFilter(PorterDuffColorFilter(Color.GREEN, PorterDuff.Mode.MULTIPLY))
        drawObject(mCarImage, canvasW - h / 2, canvasH - w / 2 - h, 90f)


        // Mark the canvas' center point
        x = canvasW / 2
        y = canvasH / 2
        val size = 2f
        trans.setTranslate(x, y)
        mPaint.color = Color.GREEN
        mCanvas.drawRect(x - size, y - size, x + size, y + size, mPaint)

        setImageBitmap(mFinalBitmap)
    }

    fun clearCanvas() {
        mCanvas.drawColor(Color.parseColor("#e5e5e5"))
        //mFinalBitmap.recycle();
        //mFinalBitmap = null;
        //invalidate();
    }

    // TODO: Remove onTouchEvent for rotating objects in the canvas.
    override fun onTouchEvent(e: MotionEvent): Boolean {
        if (e.action == MotionEvent.ACTION_DOWN) {
            oldX = e.x
            //oldY = e.getY();
        } else if (e.action == MotionEvent.ACTION_MOVE) {
            val delta = (e.x - oldX) * 0.7f
            rot += delta.toInt()
            oldX = e.x
            //oldY = e.getY();

            drawMeLikeOneOfYourIntersections()
        }
        return true
    }
}